### ISSUE TITLE
Exception handling.

### ISSUE THREAD

**Issue by Pusnow** [Solo5/solo5#327]
Hi, I'm making small OS with Solo5.
But, I have a trouble on implementing basic trap handler.
I implemented BP trap handler like below, then it cause GP exception.
Is there any tip for this?
It seems like it fails from returning from the handler. 

trap handler:
```c
void cpu_trap_handler(uint64_t num, struct trap_regs *regs)
{
    log(INFO, "Solo5: trap: type=%s ec=0x%lx rip=0x%lx rsp=0x%lx rflags=0x%lx, cs=0x%lx, ss=0x%lx\n",
        traps[num], regs->ec, regs->rip, regs->rsp, regs->rflags, regs->cs, regs->ss);

    if (num == 3){
        log(INFO, "#BP called\n");
    }
    else
    {
        PANIC("Fatal trap", regs);
    }
}
```

app_main:
```c
int solo5_app_main(const struct solo5_start_info *si __attribute__((unused)))
{
    puts("\n**** Solo5 standalone test_exception ****\n\n");

    asm("int3");

    return SOLO5_EXIT_SUCCESS;
}
```

output:

```c
\            |      ___|
  __|  _ \  |  _ \ __ \
\__ \ (   | | (   |  ) |
____/\___/ _|\___/____/
Solo5: Memory map: 512 MB addressable:
Solo5:     unused @ (0x0 - 0xfffff)
Solo5:       text @ (0x100000 - 0x105fff)
Solo5:     rodata @ (0x106000 - 0x107fff)
Solo5:       data @ (0x108000 - 0x10cfff)
Solo5:       heap >= 0x10d000 < stack < 0x20000000

**** Solo5 standalone test_exception ****

Solo5: trap: type=#BP ec=0x0 rip=0x103a00 rsp=0x1fffffe0 rflags=0x6, cs=0x8, ss=0x10
#BP called
Solo5: trap: type=#GP ec=0x10 rip=0x1018f2 rsp=0x10b038 rflags=0x10002, cs=0x8, ss=0x10
Solo5: ABORT: cpu_x86_64.c:175: Fatal trap
```

rip=0x103a00:
```asm
00000000001039e0 <solo5_app_main>:
  1039e0:   48 8d 3d 71 2f 00 00    lea    0x2f71(%rip),%rdi        # 106958 <pow10.1465+0x558>
  1039e7:   48 83 ec 08             sub    $0x8,%rsp
  1039eb:   e8 d0 fe ff ff          callq  1038c0 <strlen>
  1039f0:   48 8d 3d 61 2f 00 00    lea    0x2f61(%rip),%rdi        # 106958 <pow10.1465+0x558>
  1039f7:   48 89 c6                mov    %rax,%rsi
  1039fa:   e8 41 ca ff ff          callq  100440 <solo5_console_write>
  1039ff:   cc                      int3
  103a00:   31 c0                   xor    %eax,%eax
  103a02:   48 83 c4 08             add    $0x8,%rsp
  103a06:   c3                      retq
  103a07:   66 0f 1f 84 00 00 00    nopw   0x0(%rax,%rax,1)
  103a0e:   00 00 
```

rip= 0x1018f2
```asm
  1018cc:   41 53                   push   %r11
  1018ce:   48 c7 c7 03 00 00 00    mov    $0x3,%rdi
  1018d5:   48 89 e6                mov    %rsp,%rsi
  1018d8:   48 83 c6 48             add    $0x48,%rsi
  1018dc:   e8 5f fe ff ff          callq  101740 <cpu_trap_handler>
  1018e1:   41 5b                   pop    %r11
  1018e3:   41 5a                   pop    %r10
  1018e5:   41 59                   pop    %r9
  1018e7:   41 58                   pop    %r8
  1018e9:   59                      pop    %rcx
  1018ea:   5a                      pop    %rdx
  1018eb:   5e                      pop    %rsi
  1018ec:   5f                      pop    %rdi
  1018ed:   58                      pop    %rax
  1018ee:   48 83 c4 10             add    $0x10,%rsp
  1018f2:   48 cf                   iretq
```

**Comment by mato** [comment]
The CPU trap handlers were not written with returning from them in mind, the intent is that they report the trap and immediately abort the unikernel. This is intentional, as in our model there's no need for anything else.

To get this to work, you'd need to adjust the stack(s)/address on the return path appropriately.

**Comment by Pusnow** [comment]
I solved this issue by reloading segment registers after `lgdt`:

In `static void gdt_init(void)`
```c
    /*
     * TODO: Technically we should reload all segment registers here, in
     * practice this doesn't matter since the bootstrap GDT matches ours, for
     * now.
     */
    asm volatile("movw %%ax,%%gs" :: "a" (GDT_DESC_OFFSET(GDT_DESC_DATA)));
    asm volatile("movw %%ax,%%fs" :: "a" (GDT_DESC_OFFSET(GDT_DESC_DATA)));
    asm volatile("movw %%ax,%%es" :: "a" (GDT_DESC_OFFSET(GDT_DESC_DATA)));
    asm volatile("movw %%ax,%%ds" :: "a" (GDT_DESC_OFFSET(GDT_DESC_DATA)));
    asm volatile("movw %%ax,%%ss" :: "a" (GDT_DESC_OFFSET(GDT_DESC_DATA)));

```

By the way, the comment says "the bootstrap GDT matches ours". However, it seems they are little bit different.

bootstrap GDT:
```c
 gdt[X86_GDT_NULL] = null; // X86_GDT_NULL == 0
 gdt[X86_GDT_CODE] = sreg_to_desc(&hvt_x86_sreg_code); // X86_GDT_CODE == 1
 gdt[X86_GDT_DATA] = sreg_to_desc(&hvt_x86_sreg_data); // X86_GDT_DATA == 2
```

gdt in `gdt_init`:

```c
#define GDT_DESC_NULL           0
#define GDT_DESC_CODE           1
#define GDT_DESC_CODE32         2 /* Used by boot.S on virtio targets */
#define GDT_DESC_DATA           3
#define GDT_DESC_TSS_LO         4
#define GDT_DESC_TSS_HI         5



cpu_gdt64[GDT_DESC_CODE] = GDT_DESC_CODE_VAL; 
cpu_gdt64[GDT_DESC_DATA] = GDT_DESC_DATA_VAL;

```

There are additional space for GDT_DESC_CODE32  gdt descriptor in `cpu_gdt64`.

**Comment by mato** [comment]
> I solved this issue by reloading segment registers after lgdt:
> In static void gdt_init(void) [...]

Oh, yes, you're probably right, with the way things are currently set up we should indeed reload all segment registers after `lgdt`, otherwise we're running the (V)CPU out of spec.

> There are additional space for GDT_DESC_CODE32 gdt descriptor in cpu_gdt64.
> Is it also intended? or a bug?

There was probably a reason for it (consistency between virtio and hvt) but things are now definitely out of sync.

Thanks for the report, and the detailed diagnosis. I'll keep this issue open and write a fix in due course (currently deep in #326).
